### ISSUE TITLE
Moving beyond one dimension: adding depth to the Chakra documentation

### ISSUE THREAD

**Issue by with-heart** [chakra-ui/chakra-ui#1412]
Howdy yall! 👋 

As we get closer to the full 1.0 release, I've been allowing myself a little time to think about what the post-1.0 world looks like for Chakra. There are a lot of possibilities, but one of the most important in my mind is the need for deeper documentation.

In this issue, I'd like to present my thinking on this topic in order to (hopefully) convince you that this is a 🚨 very high priority topic 🚨 for Chakra to address. I hope that it sparks discussion within our community and helps us consider how to best improve our documentation in the future.

# Chakra as it stands

I believe that Chakra users are very diverse: they come from different backgrounds, have varying levels of skill with React, use a number of React frameworks and technologies, and are attempting to achieve a variety of different outcomes using our library.

I also believe that our website and documentation does very little to acknowledge and celebrate this diversity. Whether you're a React newbie with no understanding of component libraries, or you're an expert looking to integrate Chakra with specific frameworks or libraries like NextJS or Framer Motion, our documentation treats you as if you're the same.

The basic flow for any user coming into the Chakra world currently looks like this:

![image](https://user-images.githubusercontent.com/1954752/88504516-06449c00-cfa3-11ea-8f8b-c6d966cd4482.png)

We have a few minor guides for specific technologies and offer a number of basic starters/templates, but beyond that, our documentation is essentially a reference. It's great as a reference because it fully documents the Chakra API, but it doesn't fulfill the diverse needs of our users—far from it, in fact.

# Some examples of why this is a problem

Let's add a dimension to the situation and consider a user's skill level. Whether you're new to React, proficient at React, or a super-pro React master, here's how the flow looks for you:

![image](https://user-images.githubusercontent.com/1954752/88504529-0f356d80-cfa3-11ea-97df-0bdb7ae939f8.png)

What a React beginner needs to understand Chakra and get started with it is radically different from a React expert who already has lots of experience with building React applications.

On top of that, let's consider whether a user knows what a component library is or not! They're still a fairly new concept, so it's possible that even a React expert with many years of experience may not fully understand what that means.

![image](https://user-images.githubusercontent.com/1954752/88504412-b49c1180-cfa2-11ea-9954-1038e26e56c6.png)

With only 3 skill levels and 2 choices of whether the user understands component libraries or not, we're already at 6 paths. And as it stands, they all lead to reading the same docs!

Okay, enough with the examples. I think you can see why I view this as such a problem. What can we *do* about it?

# Personalize the documentation experience

I think we need to consider the various personas that make up our user base. These are the different combinations of diversity that make up our user base, and there are definitely **a lot** of them!

Just off the top of my head, here are some different dimensions and possibilities I can think of:

- React skill level
  - newbie
  - intermediate
  - expert
- Component library skill level
  - newbie
  - intermediate
  - expert
- Technologies/libraries to use with Chakra
  - NextJS
  - CRA
  - Parcel
  - Webpack
  - TypeScript
  - Storybook
  - Playroom
- Desired outcomes
  - Using Chakra components to build an app
  - Implementing custom themes using Chakra
  - Building and maintaining a component library on top of Chakra
  - Contributing to Chakra itself

This is far from a complete list! And it's clearly not possible for us to consider every possible combination.

But the more we try to map out common user personas and imagine the best path for them, the better our experience will be for users new to Chakra.

## Example: skill levels in React + component libraries

I threw together an example of one possible combination. You're new to React, and you either do or don't already know what a component library is:

![image](https://user-images.githubusercontent.com/1954752/88505776-625cef80-cfa6-11ea-8da2-e6feaf1d2d43.png)

This flow considers the needs of a person new to React and new to Chakra:

1. If they don't know what a component library is, provide a basic conceptual overview
1. Introduce them to the Chakra components
2. Introduce them to the Chakra theme
3. Provide an introduction tutorial that hand-holds them through creating a basic site using Chakra components and adding some simple theme modifications
4. Continue on to the docs for them to use as a reference

It provides the user with rudimentary knowledge about Chakra. Just enough to get started with building some basic things, which is what a React newbie is looking to do.

Now let's compare that same skill level breakdown, but for a user that has intermediate React skills:

![image](https://user-images.githubusercontent.com/1954752/88506041-11013000-cfa7-11ea-9c40-fee6ce7b4c7b.png)

This flow looks fairly similar to the flow for new users, except in this case, we continue on from the introduction tutorial to an overview of the component theming api, followed by a basic tutorial for theming components, which I would consider fairly intermediate topics.

It's not a huge change, but we've tailored our experience so that we give the user the knowledge they need to get started at a level they're comfortable with.

## Example: NextJS

Now let's consider the documentation from a different angle: you're interested in using NextJS with Chakra, and you either want to learn about this topic, or you just want to get started:

![image](https://user-images.githubusercontent.com/1954752/88506456-f4b1c300-cfa7-11ea-8fdb-17589e1cb3fe.png)

Here the flows are a little more diverse. We provide a user looking to learn with a few starting resources before tying them back into the "get started" flow. 

# Where to from here?

That's the hard part. Even if we had tons of contributors dedicated to this purpose, it's not an easy task. 

I think what would help would be to start with fleshing out our one-dimensional documentation. In my mind, this means we need to start with a few things:

- Beginner documentation. If the user is new to React, we should have documentation just for them. This is foundational stuff, walking the user through the very basics of Chakra and helping them to get started.
- Technology-specific documentation. Starting with bigger technologies is probably smart. If you're looking to use NextJS, we should have a collection of documentation just for that.
- A lot of the topics one-off topics listed in #811
- Other stuff that I'm too tired to prioritize right now.

Once we get some of the more foundational documentation in place, we can move on to considering more complex flows and personas and begin to craft documentation experiences.

# In conclusion

This was very much a stream-of-consciousness issue. It started with some basic thinking about how to improve our documentation that quickly led to some diagrams and trying to flesh out what I was thinking.

I'd love to hear some feedback! 

- Does this post make sense?
- What aspects stand out as good ideas? 
- Does anything seem like a bad idea?
- Does this seem doable?
- How do you think we should get started?
- Would you like to be involved?

I'd also love to hear from you if you've used our documentation and found it lacking:

- What was difficult?
- What was missing?
- What was difficult to understand about Chakra?
- What stood out as effective?

**Comment by segunadebayo** [comment]
I love the thought you've given to this @with-heart. It's amazing!

I believe we'll massively improve Chakra's documentation. I'll chew on these and see what I can come up with.

**Comment by jmiazga** [comment]
I agree that the current docs do serve as a good API reference, but it doesn't really answer the 'why'.

Why would you do something one way or another? The theme docs really lack the reasoning, it shows how you can accomplish something, but not necessarily why.

Focusing on some of the real world needs would probably help us answer they 'why' not just the 'how'.

**Comment by tomhicks** [comment]
I've been adopting and promoting Chakra at work, having been using < 1 on personal projects. My way of explaining how the theming, etc. works is something like this:

In old Chakra, and in most component libraries, you have this two-tier setup of Theme and Component, where Theme is colours and stuff, and Component is everything else.

```
+------------+
|   Theme    |
+------------+
|  Component |
+------------+
```

With new Chakra, you've got another layer in between...

```
+--------------------------+
|          Theme           |
+--------------------------+
|   Component-Theme Spec   |
+--------------------------+
| Component Implementation |
+--------------------------+
```

You are free to change or replace any of those parts, which gives you power, like this...

There is a continuum that goes between roll-everything-yourself, and an out-of-the-box, batteries-included, opinionated component library. Chakra used to sit here:

```
Artisanal CSS |--------------------------------------------------------| Component library
                                            Chakra UI (pre 1.0) ^
```

But, now it can be many things...

```
Artisanal CSS |--------------------------------------------------------| Component library
                                                                 ^ (Chakra 1.0 vanilla)
```

```
Artisanal CSS |--------------------------------------------------------| Component library
                                        ^ (Chakra 1.0 w/overridden theme spec)
```  

```
Artisanal CSS |--------------------------------------------------------| Component library
                   ^ (Chakra 1.0 just using theming system, ignoring components)
```

And the best thing is, the new system allows you to gradually migrate down than scale to wherever you need to be. If you use Material UI, you've got a Material UI app, which is great because it's quick to get up and running, but if you want to make it your own, you're looking at a rewrite.

With Chakra v1.0, you can start off using Chakra's components, then start to theme them, then start to change the theme specs, then write your own, and ultimately abandon all the Chakra components if you so wish. And you can do it incrementally, which is not to be understated - it makes it a very good choice for any project, because you get the best of both worlds...

If you don't have a designer and want to just use something out of the box, you can use Chakra. If you want to go straight away to your own specs because they've all been designed, you can use Chakra.

----

Not sure how relevant all this is, but had this concept in my head since explaining v1.0 to some people last week and wondered if it might be a helpful way to explain how and why you might want to theme things vs write your own components, etc.

**Comment by with-heart** [comment]
@tomhicks That is an excellent way to explain it! Chakra is a system of layers that allow you to vary your use case on a spectrum between the bog-standard library and your own custom designs. I love how you've laid that out so clearly with your diagrams.

That is also a direction I've been thinking in. One of the things I want to spend more time documenting is the as-of-yet undocumented tooling that we use internally to build the Chakra components:

- Component hooks & contexts
  - Most composite components have associated hooks `useX` hooks (`useModal`, `useAccordion`, etc.)
  - Used to generate a11y and display attributes, custom behaviors, event handlers, props getters, etc. that are attached to the containing component and passed to the component's `Context`
  - Context is accessed by composite components, frequently with a "props getter" fn made specifically for that part of the composite that is then spread on the part component
  - Can be used by users to remix the composite's internal behavior into a totally new system of components, or to access the composite context to add additional behavior on top of the existing
- State & handler management
  - Can automatically whether components are controlled or uncontrolled and adjust whether to use `value`+`onChange` or to manage state internally starting from the `defaultValue` state
    - Also provides warnings if component switches between controlled and uncontrolled, just like React
  - Functionality for merging `props.onX` with our internal `onX` handlers, making sure external handlers are notified of all activity
  - Functionality for merging multiple `ref` instances into one, so users are able to track element refs that we're also tracking for internal purposes
  - Can easily create many `onKeyDown` handlers by passing a `keyMap` object with handler functions for each key
- a11y management
  - Automatically manages the `id` for composite parts (derived from the `id` passed by users, if present) and establishes relationships between them when necessarily (`aria-describedBy`, `aria-labelledBy`, etc.)
  - Focus management is part of this
- Descendant management
  - Allows composite components to maintain a descendant state via `Context`
  - Useful for complex behaviors like arrow keys navigating between `AccordionItem` children by focusing each one in turn
  - Composite parts are able to easily target both the parent and siblings
  - Can read basic information about descendants like `disabled` and `focusable` state, allowing the composite to craft behaviors that adjust based on those states (like menu arrow navigation skipping descendants that aren't focusable, or prevent interactions with disabled descendants)

I would consider all of these to be part of a "component tooling" layer that lies just above the "Component Implementation" layer in your diagram. Built on top of the basic component styles layer, they allow us to build complex yet coherent components.

Additionally I see two more layers worth documenting–though this would probably be almost entirely for users looking to contribute to the system or become maintainers, as it's really getting into the guts of Chakra:

1. `@chakra-ui/styled-system`

This package is the basis for all life in the Chakra world:

- hooks into `styled-system` to establish a parser for all themed props
- creates the pseudo props and associates them with CSS
- exports the `css` function, which connects a style config object/fn with the theme to create a css object
  - applies responsive behavior to responsive props using `theme.breakpoints`
  - processes nested style configs (very important for component theming)
  - reads prop values from associated scales (e.g. `xs` -> `theme.fontSize.xs`)
  - associates props with their CSS equivalents

2. `@chakra-ui/system`

Built directly on top of the `styled-system` package. Uses the `css` function to connect the css objects with `emotion`, turning styles into actual components

- defines the functions that translate theme props, pseudo props, modifier props, etc., into css objects using the `css` function
- implements the `chakra` function
  - connects the above functions and the component's `baseStyles` with `emotion`, turning them into actual components
  - attaches TypeScript types to generated components
- creates complex types for the `as` prop to function correctly
- creates types for the `theme` and its various pieces like `baseStyles`, component theming objects, the theming-specific props that components can receive (`size`, `variant`, `colorScheme`, `orientation`), etc.
- implements `ThemeProvider`, `ChakraProvider`, `GlobalStyles`, `useTheme`

---

This comment ended up being way more detailed than I intended. Feels like it's important to get this kind of stuff out there though.

@segunadebayo Could you look over this and let me know if I'm wrong about my understanding of the underlying systems, or if I'm missing any important details?

**Comment by with-heart** [comment]
I made this awhile ago and it's now **very** outdated after many changes to the component theming system, but do yall think diagrams like this would help explain how component themes are applied?

![image](https://user-images.githubusercontent.com/1954752/88566199-b0eca700-d003-11ea-9b62-8956715efad0.png)

**Comment by segunadebayo** [comment]
I really love your thoughts and explanations @with-heart @tomhicks. I feel you both described it way better than I would ever describe 😂. Thank you for sharing 💖

@with-heart I'm happy to add these diagrams to the new docs.

@jmiazga I agree the docs need a lot of work, I'm just almost out of energy to complete these, I'd love to hand over the docs to someone else in the community and provide top-level guidance.

Thanks guys.

**Comment by DavidJFelix** [comment]
@with-heart I think that the intention to include all users and give a good starting point is great. What I'm trying to do right now is mentally wrestle the flow charts you've created into a flat documentation or at least something more normal to what people may expect when reading documentation. I think largely the flow charts you've created are *user stories* on how they would like to use the documentation, and your job as an author is to try to fit those stories.

What I've seen in the past that I think does accomplish this (although could maybe use more inter-page-linking) is to have an `Introduction` or `Getting Started` section. This section is designed to acknowledge that there may be a learning curve and the more pages you put here, the more you give to beginners to understand what Chakra is about and what component libraries can do. I've also seen people have a `Quick Start` section that is a TLDR, right to the point on how to use it. I don't necessarily think this is purely for advanced users or react experts, as lots of people have already researched and just want to try the library.

When it comes to project layouts or external library examples, I typically see documentation with specific sections focused on those frameworks. For example: 

[The fastlane documentation](https://docs.fastlane.tools/) features a section called `Best Practices` where they identify a number of different CI tools (orthogonal but related to their tool) that they know people may use and demonstrate how to best use their CI/CD of choice.

<img width="250" alt="image" src="https://user-images.githubusercontent.com/263432/88677348-28314200-d0bb-11ea-8aa7-fd67329b79f3.png">

With regards to experts, I think what is important to them is that you demonstrate designs you know to work well and patterns they might implement for advanced use cases. Typically I see this in an `Advanced` section or `Best Practices`. I think when you're writing the introduction, it's important to use multiple, clearly defined pages so experts can skim stuff like "what is a component library" to make sure their understanding matches yours, but they don't feel like the docs are overly expository. Rarely are you going to find an expert who says "the docs explained too much". Their complaint will be that the docs were hard to find what they were looking for or missed advanced use cases.

I don't necessarily like the hierarchy of the [React Native Docs](https://reactnative.dev/docs/getting-started), but one thing I do appreciate is they have links that directly describe some advanced use case that I'm looking for:

<img width="255" alt="image" src="https://user-images.githubusercontent.com/263432/88678013-fcfb2280-d0bb-11ea-9cce-d77940531faf.png">

The normal use case probably doesn't bring in external or native modules, yet there is a whole page dedicated to it that can easily be indexed by google and found by me in the sidebar when I'm trying to figure out how to use the library.


# Conclusion

I do really like the use of stories for shaping your documentation and satisfying user desires. I think it will be important for you to try to keep these stories short. For example, you might want to make a story where **an advanced user comes in with NextJS as their framework and wants to build an advanced interaction with the theme for a component library** -- instead find a way to make that 3-4 stories and give good anchor points in the documentation so people can build their own stories, restart their own stories or reenter a new story at any point. That story I identified above would be better split out as:

* An advanced user wants to know what this library does
* A NextJS user wants to know how to integrate this library correctly
* A user wants to hook into the theme
* A user is looking for design advice on a component library

Create good anchors and good pages so that google indexes their searches. Hyperlink between pages and keep the hierarchy sensible so users can discover their own questions.

**Comment by DavidJFelix** [comment]
A few docs that I think cover advanced patterns really well:

* [Reach Router](https://reach.tech/router/) - Note, they don't explicitly use "advanced" anywhere. The patterns are in "tutorials" and "guides"
* [React Router](https://reactrouter.com/web/example/preventing-transitions) - Likewise, they use "Examples" and "Guides" as top level topics for their use cases which intermix advanced and normal use cases.
* [Flutter](https://flutter.dev/docs) - This one is a bit out there, but I think because flutter is likely to be included in apps that may directly conflict or indirectly do what flutter intends to, it has some nice starting places for those advanced or irregular (its worth noting that while these aren't the NORM, flutter treats them as first class users by adding documentation for them) use cases. Specifically see sections for: "Platform Integration > C and C++ interop", "Add flutter to preexisting apps" and a few others also. I think flutter has lots of entry points and lots of exit points so this could be good to learn from.

**Comment by with-heart** [comment]
Just to note what came of this discussion:

- We have a project specifically for adding third-party integration documentation. This will mostly be post-1.0 work, as we're too busy to deal with it right now. https://github.com/chakra-ui/chakra-ui/projects/2
- Same as above except it's a project for more advanced documentation for contributors. https://github.com/chakra-ui/chakra-ui/projects/3
- I have another project in mind for improving documentation for new users. It's not created yet but will be soon, and will take a lot of inspiration from the discussion here.

**Comment by DavidJFelix** [comment]
Thanks for summarizing the action items/recap here.
